### PR TITLE
fix: improve re-opening behaviour of group updates

### LINKED ISSUE #13827: Renovate is re-creating closed group updates if not all deps have the exact same version
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

31.56.0

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

13.0.1-ee 

### Describe the bug

Renovate seems to be ignoring some closed PRs of group/monorepo updates. It just re-creates them with the next run. It is working fine for "normal" single dependency PRs. In GitLab they are marked as being immortal.

The same happens on a self-hosted Bitbucket 5.9.1.

I compared logs and it comes down to Renovate enabling `recreateClosed` for the affected group updates (see below).

Looks like this happens in case group updates consist out of multiple versions. https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/updates/generate.ts#L152
E.g. a Symfony ^5.4 -> ^6.0 update PR has deps in it consisting out of 6.0.1, 6.0.3 and so on (because Symfony is only releasing package updates if there actually were changes). I try to come up with a fix by myself, maybe the check there can be improved a bit.

https://github.com/renovatebot/renovate/discussions/10710 is describing this too.

### Relevant debug logs

The logs show that it works fine for react-monorepo but not for jest-monorepo.

<details><summary>Logs</summary>

```
[08:07:43] :	 [Step 1/1] DEBUG: getBranchPr(feature/renovate-react-monorepo) (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:43] :	 [Step 1/1] DEBUG: findPr(feature/renovate-react-monorepo, undefined, open) (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:43] :	 [Step 1/1] DEBUG: branchExists=true (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:43] :	 [Step 1/1] DEBUG: dependencyDashboardCheck=undefined (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:43] :	 [Step 1/1] DEBUG: recreateClosed is false (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:43] :	 [Step 1/1] DEBUG: findPr(feature/renovate-react-monorepo, chore(deps): update react monorepo to ~16.14.0, !open) (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:43] :	 [Step 1/1] DEBUG: Found closed PR with current title (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:43] :	 [Step 1/1] DEBUG: getPr(6) (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:43] :	 [Step 1/1] DEBUG: getMR(6) (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:44] :	 [Step 1/1] DEBUG: Closed PR already exists. Skipping branch. (repository=acme/repo, branch=feature/renovate-react-monorepo)
[08:07:44] :	 [Step 1/1]        "prTitle": "chore(deps): update react monorepo to ~16.14.0"
...
[08:07:46] :	 [Step 1/1] DEBUG: getBranchPr(feature/renovate-major-jest-monorepo) (repository=acme/repo, branch=feature/renovate-major-jest-monorepo)
[08:07:46] :	 [Step 1/1] DEBUG: findPr(feature/renovate-major-jest-monorepo, undefined, open) (repository=acme/repo, branch=feature/renovate-major-jest-monorepo)
[08:07:46] :	 [Step 1/1] DEBUG: branchExists=true (repository=acme/repo, branch=feature/renovate-major-jest-monorepo)
[08:07:46] :	 [Step 1/1] DEBUG: dependencyDashboardCheck=undefined (repository=acme/repo, branch=feature/renovate-major-jest-monorepo)
[08:07:46] :	 [Step 1/1] DEBUG: recreateClosed is true (repository=acme/repo, branch=feature/renovate-major-jest-monorepo)
[08:07:46] :	 [Step 1/1] DEBUG: Checking if PR has been edited (repository=acme/repo, branch=feature/renovate-major-jest-monorepo)
```
</details>

### Have you created a minimal reproduction repository?

https://github.com/herndlm/renovate-13827-repro/pull/2 (only after I was asked for it ;))

### PR DESCRIPTION
**Pull request by herndlm**
## Changes:

Only force re-create update if the `newValue` of all contained deps is different too. Previously only `newVersion` was used. 

## Context:

Makes group updates consisting out of different patch versions not immortal and being force re-ceated after the update was manually closed/declined.

Closes #13827

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/updates/generate.ts [hunk 1]
```diff
@@ -73,13 +73,17 @@ export function generateBranchConfig(
   const depNames: string[] = [];
   const newValue: string[] = [];
   const toVersions: string[] = [];
+  const toValues: string[] = [];
```
Comment (viceice) [inline lib/workers/repository/updates/generate.ts:76]: I think we should use a `Set<string>` here instead.

**Review by viceice — COMMENTED** [review]
Looks good so far.

File: lib/workers/repository/updates/generate.ts (hunk 1 above)
Comment (herndlm) [reply lib/workers/repository/updates/generate.ts]: adapted, should I still check via `has` or just call `add` without checking? sorry, still not that familiar with some JS things :)

File: lib/workers/repository/updates/generate.ts [hunk 2]
```diff
@@ -73,13 +73,17 @@ export function generateBranchConfig(
...
   branchUpgrades.forEach((upg) => {
     if (!depNames.includes(upg.depName)) {
       depNames.push(upg.depName);
     }
     if (!toVersions.includes(upg.newVersion)) {
       toVersions.push(upg.newVersion);
     }
+    if (!toValues.has(upg.newValue)) {
+      toValues.add(upg.newValue);
+    }
```
Comment (viceice) [inline lib/workers/repository/updates/generate.ts:86]:
```suggestion
    toValues.add(upg.newValue);
```
Set will handle duplicates faster. 🤗

File: lib/workers/repository/updates/generate.ts [hunk 3]
```diff
@@ -73,13 +73,17 @@ export function generateBranchConfig(
   const depNames: string[] = [];
   const newValue: string[] = [];
   const toVersions: string[] = [];
+  const toValues: Set<string> = new Set();
```
Comment (viceice) [inline lib/workers/repository/updates/generate.ts:76]:
```suggestion
  const toValues = new Set<string>();
```
A little bit simpler.

File: lib/workers/repository/updates/generate.ts (hunk 2 above)
Comment (herndlm) [reply lib/workers/repository/updates/generate.ts:86]: as I thought ;) thx

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**